doc/source/ray-core/objects.rst

```diff
@@ -128,41 +128,50 @@ If the current node's object store does not contain the object, the object is do
       assert(*results[1] == 1);
       assert(*results[2] == 2);
 
-Passing Objects by Reference
-----------------------------
+Passing Object Arguments
+------------------------
 
 Ray object references can be freely passed around a Ray application. This means that they can be passed as arguments to tasks, actor methods, and even stored in other objects. Objects are tracked via *distributed reference counting*, and their data is automatically freed once all references to the object are deleted.
 
-.. code-block:: python
+There are two different ways one can pass an object to a Ray task or method. Depending on the way an object is passed, Ray will decide whether to *de-reference* the object prior to task execution.
 
-    @ray.remote
-    def echo(x):
-        print(x)
+**Passing an object as a top-level argmuent**: When an object is passed directly as a top-level argument to a task, Ray will de-reference the object. This means that Ray will fetch the underlying data for all top-level object reference arguments, not executing the task until the object data becomes fully available.
 
-    # Put an object in Ray's object store.
-    object_ref = ray.put(1)
+.. literalinclude:: doc_code/obj_val.py
 
-    # Pass-by-value: send the object to a task as a top-level argument.
-    # The object will be de-referenced, so the task only sees its value.
-    echo.remote(object_ref)
-    # -> prints "1"
+**Passing an object as a nested argument**: When an object is passed within a nested object, for example, within a Python list, Ray will *not* de-reference it. This means that the task will need to call ``ray.get()`` on the reference to fetch the concrete value. However, if the task never calls ``ray.get()``, then the object value never needs to be transferred to the machine the task is running on.
 
-    # Pass-by-reference: when passed inside a Python list or other data structure,
-    # the object ref is preserved. The object data is not transferred to the worker
-    # when it is passed by reference, until ray.get() is called on the reference.
-    echo.remote({"obj": object_ref})
-    # -> prints "{"obj": ObjectRef(...)}"
+.. literalinclude:: doc_code/obj_ref.py
 
-    # Objects can be nested within each other. Ray will keep the inner object
-    # alive via reference counting until all outer object references are deleted.
-    object_ref_2 = ray.put([object_ref])
+We recommend passing objects as top-level arguments where possible, but nested arguments can be useful for passing objects on to other tasks without needing to see the data.
```

Comment:
I think it would be better to put this right after the description of the first method so it's easier to see.